src/Access/ImplicitGrants.cpp

```diff
@@ -0,0 +1,138 @@
+        static const AccessFlags alter_materialize_ttl = AccessType::ALTER_MATERIALIZE_TTL;
+        if (res & alter_ttl)
+            res |= alter_materialize_ttl;
+
+        /// RELOAD_DICTIONARY (global) => RELOAD_EMBEDDED_DICTIONARIES (global)
+        static const AccessFlags reload_dictionary = AccessType::SYSTEM_RELOAD_DICTIONARY;
+        static const AccessFlags reload_embedded_dictionaries = AccessType::SYSTEM_RELOAD_EMBEDDED_DICTIONARIES;
+        if ((level == 0) && (min_flags_with_children & reload_dictionary))
+            res |= reload_embedded_dictionaries;
+
+        /// any column flag => SHOW_COLUMNS => SHOW_TABLES => SHOW_DATABASES
+        ///                  any table flag => SHOW_TABLES => SHOW_DATABASES
+        ///       any dictionary flag => SHOW_DICTIONARIES => SHOW_DATABASES
+        ///                              any database flag => SHOW_DATABASES
+        static const AccessFlags show_columns = AccessType::SHOW_COLUMNS;
+        static const AccessFlags show_tables = AccessType::SHOW_TABLES;
+        static const AccessFlags show_dictionaries = AccessType::SHOW_DICTIONARIES;
+        static const AccessFlags show_tables_or_dictionaries = show_tables | show_dictionaries;
+        static const AccessFlags show_databases = AccessType::SHOW_DATABASES;
+
+        if (res & AccessFlags::allColumnFlags())
+            res |= show_columns;
+
+        if ((res & AccessFlags::allTableFlags())
+            || (level <= 2 && (res & show_columns))
+            || (level == 2 && (max_flags_with_children & show_columns)))
+        {
+            res |= show_tables;
+        }
+
+        if (res & AccessFlags::allDictionaryFlags())
+            res |= show_dictionaries;
+
+        if ((res & AccessFlags::allDatabaseFlags())
+            || (level <= 1 && (res & show_tables_or_dictionaries))
+            || (level == 1 && (max_flags_with_children & show_tables_or_dictionaries)))
+        {
+            res |= show_databases;
+        }
+
+        return res;
+    };
+
+    AccessRights res = access;
+    res.modifyFlags(modifier);
+
+    res.grant(getConstImplicitGrants());
+    return res;
+}
+
+const AccessRightsElements & getConstImplicitGrants()
+{
+    static const AccessRightsElements elements = []
+    {
+        AccessRightsElements res;
+
+        /// System tables any user can SELECT from.
+        AccessFlags select_flags = AccessType::SELECT;
+        constexpr std::array tables_in_system_db = {
+            "one",
+            "numbers",
+            "numbers_mt",
+            "zeros",
+            "zeros_mt",
+            "functions",
+            "settings",
+            "formats",
+            "table_functions",
+            "aggregate_function_combinators",
+            "data_type_families",
+            "collations",
+            "table_engines",
+            "contributors",
+            "privileges",
+            "licenses",
+            "time_zones",
```

Comment:
@alexey-milovidov Please take a look at this list, maybe you suggest to add/remove something here. This is a list of tables in the system database with implicitly granted access. That means any user will have access to those tables without executing `GRANT` statement. Before this PR access to all the system database was implicitly granted and this was undesirable for tables like `system.zookeeper`, `system.clusters`, or custom tables.